### ISSUE TITLE
Types created by a function not allowed in type expression

### ISSUE THREAD

**Issue by mauvilsa** [microsoft/pyright#6000]
**Describe the bug**
Types created via a function with `type` return, when used as type annotation incorrectly gives `Variable not allowed in type expression Pylance reportGeneralTypeIssues`. A `type` return clearly states that the function returns a type. The normal way to use such a type in a type annotation, is to call this function and assign the return to a variable. Then use this variable as the annotation. There should be no error.

**Code or Screenshots**

```python
from typing import Optional

def create_custom_type(name: str, docstring: Optional[str] = None) -> type:
    class CustomType:
        pass
    return type(name, (CustomType,), {"__doc__": docstring})

my_type = create_custom_type("my_type", "my_type docstring")

def my_function(arg: my_type):
    pass
```

![Screenshot from 2023-09-21 21-59-48](https://github.com/microsoft/pyright/assets/5780272/306b56fc-11db-420c-9212-b8e6c37e726e)

**VS Code extension or command-line**
VS Code extension v2023.9.20

**Comment by erictraut** [comment]
Yes, this behavior is correct. Variables are not allowed within a type expression.

In your code, you're dynamically creating a new type and assigning it to a variable `my_type`. A static type checker cannot reason about dynamically-created types beyond the fact that it's a subclass of `type`. You can therefore change the type annotation for `arg` to `type`. 

```python
def my_function(arg: type):
    pass
```

**Comment by mauvilsa** [comment]
But, a dynamically created type is quite similar to a class. And a static type checker can be more helpful. Take for instance:

```python
def my_function(arg: my_type = my_type("value being casted")):  # all good
    pass

my_var = my_type("some other value")

my_function(my_var)  # all good
my_function(0)  # unexpected argument
```

There isn't much difference if `my_type` is returned by a function with `type` return, or it is `class my_type: ...`.

**Comment by erictraut** [comment]
Variables are not allowed in static type annotations in the Python type system, nor should they be. This is enforced not only by pyright but other type checkers like mypy. Pyright is correctly following the type system by enforcing this limitation.

> a dynamically created type is quite similar to a class

Yes, from the perspective of static type checking, the type of a dynamically-created type is simply `type` (or `type[Any]`), since a class is always an instance of `type`, and there's not much more you can say about it. So you might as well use `type` (or `type[Any]`) as the type annotation for `arg`.
